src/cargo/core/compiler/fingerprint.rs

```diff
@@ -1866,10 +1867,53 @@ fn find_stale_file(
     None
 }
 
+type Svh = String;
+fn get_svh_from_ar<R: Read>(reader: R) -> Option<Svh> {
+    //    use std::fs::File;
+    //    let file = File::open(path).ok()?;
+
+    let mut ar = ar::Archive::new(reader);
+    while let Some(file) = ar.next_entry() {
+        if let Ok(file) = file {
+            let s = String::from_utf8_lossy(&file.header().identifier());
+            if s.ends_with(".rmeta") {
+                if let Some(index) = s.rfind('-') {
+                    return Some(s[index + 1..(s.len() - ".rmeta".len())].to_string());
+                }
+            }
+        }
+    }
+    None
+}
+
+fn get_svh_from_object_file<R: Read>(mut reader: R) -> Option<Svh> {
+    use object::Object;
+    //    use std::fs::File;
+    //   use std::io::Read;
+
+    //let mut file = File::open(path).ok()?;
+    let mut data = vec![];
+    reader.read_to_end(&mut data).ok()?; //TODO: looks expensive!
+    let obj = object::read::File::parse(&data).ok()?;
+
+    for (_idx, sym) in obj.symbols() {
+        //TODO: symbol is at the end typically.
+        if let Some(name) = sym.name() {
+            if name.starts_with("_rust_svh_") {
+                if let Some(index) = name.rfind('_') {
+                    return Some(name[index + 1..].to_string());
+                }
+            }
+        }
+    }
+    None
+}
+
 #[derive(Clone, Copy, Eq, PartialEq)]
 pub enum FileHashAlgorithm {
-    Md5,
+    /// Svh is embedded as a symbol or for rmeta is in the .rmeta filename.
```

Comment:
The `.rmeta` filename contains the `-Cextra-filename` (which matches `-Cmetadata` for cargo), not SVH. The SVH isn't know until late during the compilation. The `-Cextra-filename` is calculated by cargo using everything except for the source code. It is meant to stay stable when changing the source code. Only changes to a `Cargo.toml`/`Cargo.lock` or the selected profile should affect it.

Reply:
Interesting - I was wondering what the extra-filename was used for.
Will update the comment to make it clear that the svh is in the .rmeta filename inside an .rlib.

Reply:
The extra-filename is used to prevent conflicts between the filenames of multiple versions of the same crate. The metadata is used to prevent conflicts between symbols of multiple versions of the same crate.